검색 기능 추가 구현(Rooms)

Mulgyeol opened the pull request:
## 개요
검색창에서 Room 타입 카테고리를 선택 후 검색하면 키워드가 제목에 포함된 방들이 보여지도록 구현했습니다.

## 작업사항
- room 검색 API 작성
- room service에서 키워드가 포함된 방 검색
- search view에서 검색된 방 RoomCard를 보여주도록 구현

## 변경로직
- 기존 EventsList를 보여주는 로직을 참고해서 구현해봤습니다.
- loading이 아니라면 event리스트를 보여주던 코드를 itemList함수 형태로 분리했습니다.

### 변경후

https://user-images.githubusercontent.com/59464537/141655407-0a33842e-76dc-4d80-843f-76c5c2d1ae0b.mov


### 에러 기록
- 타입을 선택하고(Rooms, Events) 검색을 한 이후에 타입을 바꾸면 에러가 발생하는 상황입니다.


https://user-images.githubusercontent.com/59464537/141655371-5ad20b17-02bc-422f-bfb2-4e8e4bf76d37.mov

Mulgyeol committed:
♻️ 사용되지 않는 변수 삭제

const result 삭제 후 insert만 수행하도록 수정

Mulgyeol committed:
♻️ 불필요한 공백 삭제

Mulgyeol committed:
✨ Room 검색 API 구현(server)

Mulgyeol committed:
✨ Room 검색 구현

keyword가 포함된 room을 검색해서 RoomCard를 보여주는 기능 구현

iHoHyeon commented on client/src/views/search-view.tsx:
```diff
@@ -82,6 +85,28 @@ function SearchView() {
     }
   }, []);
 
+  const setRoomView = useSetRecoilState(roomViewType);
+  const setRoomDocumentId = useSetRecoilState(roomDocumentIdState);
+
+  const roomCardClickHandler = (e: MouseEvent) => {
+    const RoomCardDiv = (e.target as HTMLDivElement).closest('.RoomCard');
+    const roomDocumentId = RoomCardDiv?.getAttribute('data-id');
+    setRoomView('inRoomView');
+    if (roomDocumentId) setRoomDocumentId(roomDocumentId);
+    else console.error('no room-id');
+  };
+
```
event 부분 작성하면서 느꼈던 부분인데, 이 부분에서 결국 view 컴포넌트에서 필요한건 `roomCardClikcHandler`이므로 이것만 반환해주는 커스텀훅으로 바깥으로 뺄 수 있을 것 같아요! room-view와 search-view에서 공통으로 사용되는 부분이기도 하고 여러 상태를 search-view 컴포넌트에 작성하려고 하니 계속 길어지더라구요..

iHoHyeon commented on client/src/views/search-view.tsx:
```diff
@@ -82,6 +85,28 @@ function SearchView() {
+  };
+
+  // eslint-disable-next-line consistent-return
+  const itemList = () => {
+    if (searchType === 'Events') {
+      return <EventCardList setEventModal={setEventModal} eventList={nowItemsList} />;
+    }
+
+    if (searchType === 'Rooms') {
+      return <RoomCardList roomCardClickHandler={roomCardClickHandler} roomList={nowItemsList} />;
+    }
+  };
+
```
searchType에 따라서 어떻게 view를 컨트롤할지 고민이었는데 좋은 방법인 것 같습니다 ㅎㅎ

iHoHyeon replied:
나중엔 switch로 바꿀 수 있을 것 같은데 if else 랑 switch 중에 어떤 형식이 좋을까요? 아니면 List 컴포넌트와 searchType을 Mapping 해놓아도 좋을 것 같네요

Mulgyeol replied:
임시적으로 구현해봤는데, if-else 구문은 린트가 알아서 if로만 구성하도록 바꿔주더라구요! 가독성을 고려하면 switch가 더 좋을 것 같기도 합니다.  Top 기능까지 고려하면 List응답으로 searchType을 반환해서 구분하는 것도 좋을 것 같아요!

iHoHyeon commented on server/src/api/routes/search.ts:
```diff
@@ -11,10 +12,35 @@ export default (app: Router) => {
     if (typeof keyword !== 'string' || typeof count !== 'string') {
-        res.json({ ok: false });
+      res.json({ ok: false });
     } else {
-        const items = (await eventsService.searchEvent(keyword, Number(count)))?.map(eventsService.makeItemToEventInterface);
-        res.json({ ok: true, items, keyword })
+      const items = (
+        await eventsService
+          .searchEvent(keyword, Number(count)))
+        ?.map(eventsService.makeItemToEventInterface);
+      res.json({ ok: true, items, keyword });
+    }
+  });
```
더러운 코드를 거두어주셔서 감사합니다.. ㅜ

iHoHyeon commented on server/src/services/rooms-service.ts:
```diff
@@ -58,6 +58,29 @@ class RoomService {
   }
+
+  // eslint-disable-next-line consistent-return
+  async searchRooms(keyword: string, count: number) {
+    try {
+      const query = new RegExp(keyword, 'i');
+      const res = await Rooms.find({ title: query }).sort({ date: 1 }).skip(count).limit(10);
+
+      const roomsInfo = await Promise.all((res).map(async ({
+        _id, title, isAnonymous, participants,
+      }) => {
+        const userList = participants.map(({ userDocumentId }) => ({ _id: userDocumentId }));
+        const participantsInfo = await Users.find({ _id: { $in: userList } }, ['userName', 'profileUrl']);
+        const roomInfo = {
+          _id, title, isAnonymous, participantsInfo,
+        };
+        return roomInfo;
+      }));
+
```
room을 조회하는 부분을 참고하신 것 같네요! 저번에 찬희님과 같이 작성했던 부분인데 리팩토링을 진행하지 못했었습니다ㅜ 나중에 리팩토링에서 map 내부를 함수로 빼면 가독성이 좋아질 것 같아요

Mulgyeol replied:
맞아요.. event 부분에서 작성해주신 makeItemToEventInterface 처럼 함수를 따로 빼보려고 시도했는데, 사용되는 로직이 많기도 하고 오류가 나서 일단 그대로 사용했습니다. 리팩토링이 필요한 부분입니다..!

iHoHyeon commented on server/src/services/users-service.ts:
```diff
@@ -67,7 +67,7 @@ class UserService {
   async signup(info: ISignupUserInfo) {
     try {
-      const result = await Users.insertMany([info]);
+      await Users.insertMany([info]);
```
별 건 아니지만 저번에 save() 를 call 하지 않기위해서 inserMany로 바꿨던 걸로 기억하는데 create와 insertMany의 차이점에 대해서 관련된 자료가 있더라구요.
https://masteringjs.io/tutorials/mongoose/create

NEM-NE replied:
아 insertMany는 다량의 데이터를 넣을 때 성능이 좋지만 한개의 데이터만 넣을 때는 create가 더 좋다는 뜻이군요! 그렇다면 create와 save를 동시에 해주는 insertOne이라는 유틸 함수를 만들어서 관리를 해주면 좋을 듯 싶습니다.

iHoHyeon reviewed (commented):
수고하셨습니다 ㅎㅎ 깔끔하게 작성해주셔서 리뷰하기 좋았습니다!

에러 관련해서는 카테고리 변경 시 nowItemsList가 초기화되지 않고 Type에 맞는 새로운 List를 받아오지 않아서 발생하는 에러 같습니다. 
카테고리 클릭 시 nowItemsList 를 reset해주고 nowFetching을 변경해주어서 새로운 List를 받아오게한다면 해결되지 않을까 생각해봅니다!

Mulgyeol committed:
🐛 searchType을 변경하면 리스트타입이 맞지않아 생기는 오류 해결

1. useEffect : restItemList()와 setNowFetching을 통해 초기화 싴며주는 부분의 로직이 searchRequestHandler의 로직과 일치하는 부분이 있어 searchRequestHandler 호출 한 번으로 변경해 주었습니다. (피드백 필요)
2. itemList에서 현재 검색한 아이템의 옵션과 searchType이 일치하지 않는 경우 LoadingSpinner를 반환하도록 추가해주었습니다.(리팩토링 필요)

Mulgyeol commented:
## 개요
🐛 searchType을 변경하면 리스트타입이 맞지않아 생기는 오류 해결


## 작업사항
- room 검색 API 작성
- room service에서 키워드가 포함된 방 검색
- search view에서 검색된 방 RoomCard를 보여주도록 구현

## 변경로직

1. useEffect : restItemList()와 setNowFetching을 통해 초기화 시켜주는 부분의 로직이 searchRequestHandler의 로직과 일치하는 부분이 있었습니다. 또,  searchInfo.option 값의 변경이 필요하기 때문에 searchRequestHandler 호출 한 번으로 변경해 주었습니다.
```ts
//before

useEffect(() => {
  resetItemList();
  setNowFetching(true);

  return () => resetItemList();
}, []);


//after

useEffect(() => {
  searchRequestHandler();
}, [searchType]);
```

2. itemList에서 현재 검색한 아이템의 옵션과 searchType이 일치하지 않는 경우 LoadingSpinner를 반환하도록 추가해주었습니다.(리팩토링 필요)
```ts
// before
  const itemList = () => {
    if (searchType === 'Events') {
      return <EventCardList setEventModal={setEventModal} eventList={nowItemsList} />;
    }

    if (searchType === 'Rooms') {
      return <RoomCardList roomCardClickHandler={roomCardClickHandler} roomList={nowItemsList} />;
    }
  };

//after
  const itemList = () => {
    if (searchInfo.current.option !== searchType.toLocaleLowerCase()) {
      return <LoadingSpinner />;
    }

    if (searchType === 'Events') {
      return <EventCardList setEventModal={setEventModal} eventList={nowItemsList} />;
    }

    if (searchType === 'Rooms') {
      return <RoomCardList roomCardClickHandler={roomCardClickHandler} roomList={nowItemsList} />;
    }
  };
```



### 변경 후 동작화면

https://user-images.githubusercontent.com/59464537/141677481-795d4f19-0bb4-4e80-8910-8288aa958915.mp4


### 삽질 기록
- useResetRecoilState : 아래와 같은 방식으로 searchType(카테고리를 선택했을 때 변경되는 상태)의 변경이 일어나면 item을 reset하고 fetching 상태를 변경해주려고 했습니다. 그러나, resetItemList이후 nowItemList를 출력해봤더니 itemList의 값들이 변경되지 않았습니다.
```ts
useEffect(() => {
    resetItemList();
    setNowFetching(true);
  }, [searchType]);

```
- setTimeOut으로 1초의 텀을 주고 출력을 해도 변경되지 않았는데, 원인은 searchType이 아니라 searchInfo.option에 맞춰서 fetch 요청하기 때문이었습니다..😅😅(제가 작성을 해놓고도, 흐름이 당연히 searchType이겠거니 해서 시간을 날렸네요..) 아래와 같은 searchRequestHandler 안의 로직이 적용되어야 데이터의 변경이 제대로 일어날 수 있었습니다.

```
    searchInfo.current.option = searchType.toLocaleLowerCase();
```

NEM-NE commented on server/src/api/routes/search.ts:
```diff
@@ -11,10 +12,35 @@ export default (app: Router) => {
     const { count } = req.query;
     if (typeof keyword !== 'string' || typeof count !== 'string') {
-        res.json({ ok: false });
+      res.json({ ok: false });
```
지난번 로그인, 회원가입 api에서도 ok라는 걸 사용하는 걸 봤는데 다른 API에서도 통일을 할까요?

NEM-NE commented on server/src/api/routes/search.ts:
```diff
@@ -11,10 +12,35 @@ export default (app: Router) => {
+  //   }
+  // });
+
+  searchRouter.get('/rooms/:keyword', async (req: Request, res: Response) => {
```
노션 API 명세에서 해당 관련 명세가 없는 거 같은데 어떤 내용을 담고 있나요?

NEM-NE commented on server/src/services/rooms-service.ts:
```diff
@@ -58,6 +58,28 @@ class RoomService {
+  async searchRooms(keyword: string, count: number) {
+    try {
+      const query = new RegExp(keyword, 'i');
+      const res = await Rooms.find({ $or: [{ title: query }, { description: query }] }).sort({ date: 1 }).skip(count).limit(10);
```
정규식으로 찾는 방식인가요? 신기하네요 ㅎㅎ

NEM-NE reviewed (commented):
주말동안 고생하셨습니다!

실제로 어떻게 돌아가는지 확인하고 싶었네요 ㅎㅎ 검색 기능은 직접 맡지 않아서 아직 이해가지 않은 부분이 조금 있는데요 해당 부분은 내일 알려주시면 감사하겠습니다!

NEM-NE commented on client/src/views/search-view.tsx:
```diff
@@ -41,33 +44,30 @@ function SearchView() {
-  };
-
   useEffect(() => {
     if (nowItemsList && (nowItemTypeRef.current === 'recent' || nowItemTypeRef.current === inputKeywordRef.current?.value)) {
```
(nowItemTypeRef.current === 'recent' || nowItemTypeRef.current === inputKeywordRef.current?.value)을 추가해준 이유가 있을까요?

NEM-NE commented on client/src/views/search-view.tsx:
```diff
@@ -82,6 +82,32 @@ function SearchView() {
+
+  // eslint-disable-next-line consistent-return
+  const itemList = () => {
+    if (searchInfo.current.option !== searchType.toLocaleLowerCase()) {
+      return <LoadingSpinner />;
+    }
```
if문을 제거하고 맨 아래에 둔다면 모든 예외를 처리할 수 있어서 좋지 않을까요?
